File: app/schemas/story.py
from app.models import StoryStatus
from pydantic import BaseModel
from typing import List, Optional
from datetime import datetime
from app.schemas.chapter import ChapterListItem
from app.schemas.target import TargetResponse

class CreateStoryRequest(BaseModel):
    title: str

class UpdateStoryRequest(BaseModel):
    title: Optional[str] = None
    status: Optional[StoryStatus] = None

class StoryCardResponse(BaseModel):
    id: str
    latest_chapter_id: Optional[str] = None
    title: str
    status: StoryStatus
    total_chapters: int
    word_count: int
    created_at: datetime
    updated_at: datetime

class StoryListItemResponse(BaseModel):
    id: str
    title: str
    word_count: int
    targets: List[TargetResponse]

class StoryDetailResponse(StoryCardResponse):
    chapters: List['ChapterListItem']

class StoryGridResponse(BaseModel):
    stories: List['StoryCardResponse']

File: frontend/src/infrastructure/api/types/story.ts
import { z } from "zod"
import { DateTimeSchema, StoryStatusSchema } from "./common"
import { ChapterListItemSchema } from "./chapter"

// ─── Requests ────────────────────────────────────────────────

export const CreateStoryRequestSchema = z.object({
    title: z.string().min(1).max(255),
})
export type CreateStoryRequest = z.infer<typeof CreateStoryRequestSchema>

export const UpdateStoryRequestSchema = z.object({
    title: z.string().min(1).max(255).optional(),
    status: StoryStatusSchema.optional(),
})
export type UpdateStoryRequest = z.infer<typeof UpdateStoryRequestSchema>

// ─── Responses ───────────────────────────────────────────────

export const StoryCardResponseSchema = z.object({
    storyId: z.string(),
    status: StoryStatusSchema,
    chapterNumber: z.int(),
    title: z.string(),
    wordCount: z.int().default(0),
    updatedAt: DateTimeSchema
})
export type StoryCardResponse = z.infer<typeof StoryCardResponseSchema>

export const StoryDetailResponseSchema = StoryCardResponseSchema.extend({
    chapters: z.array(ChapterListItemSchema),
})
export type StoryDetailResponse = z.infer<typeof StoryDetailResponseSchema>

export const StoryGridResponseSchema = z.object({
    stories: z.array(StoryCardResponseSchema),
})
export type StoryGridResponse = z.infer<typeof StoryGridResponseSchema>

export const PulseDimensionSchema = z.object({
    label: z.enum(["healthy", "watch", "needs-attention", "unavailable"]),
    headline: z.string(),
    report: z.string()
})
export const BookPulseResponseSchema = z.object({
    characters: PulseDimensionSchema,
    plot: PulseDimensionSchema,
    structure: PulseDimensionSchema,
    world: PulseDimensionSchema
})
export type BookPulseResponse = z.infer<typeof BookPulseResponseSchema>